metrics/fds_linux.go

```diff
@@ -4,10 +4,109 @@
 
 package metrics
 
-import "os"
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"syscall"
+	"unsafe"
+)
+
+var procSelfFDName = []byte("/proc/self/fd\x00")
 
 func currentFDs() int {
-	// TODO(bradfitz): do this without so many allocations on Linux.
-	ents, _ := os.ReadDir("/proc/self/fd")
-	return len(ents)
+	r0, _, e1 := syscall.Syscall(syscall.SYS_OPEN,
+		uintptr(unsafe.Pointer(&procSelfFDName[0])),
+		0, 0)
+	if e1 != 0 {
+		return 0
+	}
+	fd := int(r0)
+	defer syscall.Close(fd)
+
+	count := 0
+
+	const blockSize = 8 << 10
+	buf := make([]byte, blockSize) // stack-allocated; doesn't escape
+	bufp := 0                      // starting read position in buf
+	nbuf := 0                      // end valid data in buf
+	dirent := &syscall.Dirent{}
+	var err error
+	for {
+		if bufp >= nbuf {
+			bufp = 0
+			nbuf, err = readDirent(fd, buf)
+			if err != nil {
+				log.Printf("currentFDs: readDirent: %v", err)
+				return 0
+			}
+			if nbuf <= 0 {
+				return count
+			}
+		}
+		consumed, name := parseDirEnt(dirent, buf[bufp:nbuf])
+		bufp += consumed
+		if len(name) == 0 || string(name) == "." || string(name) == ".." {
+			continue
+		}
+		count++
+	}
+}
+
+func direntNamlen(dirent *syscall.Dirent) uint64 {
+	const fixedHdr = uint16(unsafe.Offsetof(syscall.Dirent{}.Name))
+	nameBuf := (*[unsafe.Sizeof(dirent.Name)]byte)(unsafe.Pointer(&dirent.Name[0]))
+	const nameBufLen = uint16(len(nameBuf))
+	limit := dirent.Reclen - fixedHdr
+	if limit > nameBufLen {
+		limit = nameBufLen
+	}
+	nameLen := bytes.IndexByte(nameBuf[:limit], 0)
+	if nameLen < 0 {
+		panic("failed to find terminating 0 byte in dirent")
+	}
+	return uint64(nameLen)
+}
+
+func parseDirEnt(dirent *syscall.Dirent, buf []byte) (consumed int, name []byte) {
```

Comment:
Can this stuff go upstream in x/sys?

Reply:
It is, but it allocs a bunch there.

Reply:
I mean, can you upstream a new variant that doesn't alloc?

If we can't justify it upstream, can we justify carrying it here? Are these allocs important?

Reply:
Started the process: https://github.com/golang/go/issues/48161

trunkd numbers says it's important